File: apps/server/src/crypto.js
// 信封加密（Envelope Encryption）三层结构 —— 对齐设计文档 §3.2
//   Root Key (KEK, 主密钥)  <- 文件 / 环境变量（KMS 适配层为 v1 后续迭代）
//     └─ Org Data Key (DEK, 每组织一把)  <- 加密后存 DB，使用时惰性解密入内存
//        └─ Secret Value (AES-256-GCM)  <- DB 只存密文 + nonce
// 算法：AES-256-GCM（认证加密），nonce 12 字节随机，绝不复用。
// 注意：本 MVP 用 Node.js 标准库实现以快速验证设计；Go 版将改用 Argon2id（密码哈希）
// 与 google/tink 风格封装，设计契约不变。
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

const KEY_LEN = 32;
const NONCE_LEN = 12;

export function loadMasterKey(dataDir) {
  // 1) 环境变量优先：TABOO_MASTER_KEY，支持 base64 或 hex（32 字节）
  const env = process.env.TABOO_MASTER_KEY;
  if (env) {
    const buf = /^[0-9a-f]{64}$/i.test(env)
      ? Buffer.from(env, 'hex')
      : Buffer.from(env, 'base64');
    if (buf.length !== KEY_LEN) throw new Error('TABOO_MASTER_KEY must be 32 bytes (hex or base64)');
    return buf;
  }
  // 2) 本地文件模式：首次启动自动生成（0600 权限），对标 Vault 的 unseal key 简化版
  fs.mkdirSync(dataDir, { recursive: true });
  const keyFile = path.join(dataDir, 'master.key');
  if (!fs.existsSync(keyFile)) {
    const key = crypto.randomBytes(KEY_LEN);
    fs.writeFileSync(keyFile, key.toString('hex'), { mode: 0o600 });
    console.log(`[taboo] master key generated at ${keyFile} (keep it safe, back it up)`);
  }
  const raw = fs.readFileSync(keyFile, 'utf8').trim();
  const buf = Buffer.from(raw, 'hex');
  if (buf.length !== KEY_LEN) throw new Error(`invalid master key file: ${keyFile}`);
  return buf;
}

export function generateDEK() {
  return crypto.randomBytes(KEY_LEN);
}

// ---- AES-256-GCM 加解密（KEK 包 DEK、DEK 包明文，共用同一原语） ----

export function encrypt(masterKey, plaintext) {
  const nonce = crypto.randomBytes(NONCE_LEN);
  const cipher = crypto.createCipheriv('aes-256-gcm', masterKey, nonce);
  const ct = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  // 存储格式: nonce | tag | ciphertext  (base64)
  return Buffer.concat([nonce, tag, ct]).toString('base64');
}

export function decrypt(masterKey, payload) {
  const buf = Buffer.from(payload, 'base64');
  const nonce = buf.subarray(0, NONCE_LEN);
  const tag = buf.subarray(NONCE_LEN, NONCE_LEN + 16);
  const ct = buf.subarray(NONCE_LEN + 16);
  const decipher = crypto.createDecipheriv('aes-256-gcm', masterKey, nonce);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ct), decipher.final()]).toString('utf8');
}

// ---- 组织 DEK 惰性缓存 ----
const dekCache = new Map(); // orgId -> Buffer(32)

export function getOrgDEK(masterKey, org) {
  let dek = dekCache.get(org.id);
  if (!dek) {
    dek = Buffer.from(decrypt(masterKey, org.dek_encrypted), 'hex');
    dekCache.set(org.id, dek);
  }
  return dek;
}

// ---- 密码哈希：MVP 用内置 scrypt；Go 版迁移至 Argon2id(m=64MB,t=3,p=4) ----

const SCRYPT_PARAMS = { N: 16384, r: 8, p: 1 }; // MVP 参数，生产环境上调

export function hashPassword(password) {
  const salt = crypto.randomBytes(16);
  const hash = crypto.scryptSync(password, salt, 64, SCRYPT_PARAMS);
  return `scrypt$${SCRYPT_PARAMS.N}$${SCRYPT_PARAMS.r}$${SCRYPT_PARAMS.p}$${salt.toString('hex')}$${hash.toString('hex')}`;
}

export function verifyPassword(password, stored) {
  const [algo, N, r, p, saltHex, hashHex] = String(stored).split('$');
  if (algo !== 'scrypt') return false;
  const hash = crypto.scryptSync(password, Buffer.from(saltHex, 'hex'), 64, {
    N: Number(N), r: Number(r), p: Number(p),
  });
  return crypto.timingSafeEqual(hash, Buffer.from(hashHex, 'hex'));
}

// ---- JWT (HS256) —— MVP 实现，TTL 15min ----

const b64u = (buf) => Buffer.from(buf).toString('base64url');

export function signJWT(payload, secret, ttlSec = 900) {
  const header = b64u(JSON.stringify({ alg: 'HS256', typ: 'JWT' }));
  const body = b64u(JSON.stringify({ ...payload, exp: Math.floor(Date.now() / 1000) + ttlSec }));
  const sig = crypto.createHmac('sha256', secret).update(`${header}.${body}`).digest('base64url');
  return `${header}.${body}.${sig}`;
}

export function verifyJWT(token, secret) {
  const [header, body, sig] = String(token).split('.');
  if (!header || !body || !sig) return null;
  const expect = crypto.createHmac('sha256', secret).update(`${header}.${body}`).digest();
  const got = Buffer.from(sig, 'base64url');
  if (expect.length !== got.length || !crypto.timingSafeEqual(expect, got)) return null;
  try {
    const payload = JSON.parse(Buffer.from(body, 'base64url').toString());
    if (!payload.exp || payload.exp < Math.floor(Date.now() / 1000)) return null;
    return payload;
  } catch {
    return null;
  }
}

export function sha256(input) {
  return crypto.createHash('sha256').update(input).digest('hex');
}

export const randomId = () => crypto.randomUUID();

File: apps/server-go/internal/crypto/crypto.go
// 信封加密（Envelope Encryption）+ 密码哈希 + JWT —— 对齐设计文档 §3.2 / §8
//
//	Root Key (KEK, 主密钥)        文件/环境变量/KMS（KMS 适配层后续迭代）
//	  └─ Org Data Key (DEK)        加密后存 DB，惰性解密入内存
//	     └─ Secret Value           AES-256-GCM，nonce 12B 随机绝不复用
//
// 密码哈希：Argon2id（m=64MB, t=3, p=4）；保留 scrypt 校验路径，
// 登录成功后透明重哈希完成迁移（issue #2 验收项）。
// JWT：HS256，TTL 15min。
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"

	"golang.org/x/crypto/argon2"
	"golang.org/x/crypto/scrypt"
)

const (
	KeyLen   = 32
	NonceLen = 12
)

// ---------- Root Key ----------

func ParseMasterKey(s string) ([]byte, error) {
	s = strings.TrimSpace(s)
	if b, err := hex.DecodeString(s); err == nil && len(b) == KeyLen {
		return b, nil
	}
	if b, err := base64.StdEncoding.DecodeString(s); err == nil && len(b) == KeyLen {
		return b, nil
	}
	return nil, errors.New("master key must be 32 bytes (hex or base64)")
}

// ---------- AES-256-GCM 原语（KEK 包 DEK、DEK 包明文共用） ----------

// Encrypt 输出: nonce(12) | tag(16) | ciphertext，base64 编码
func Encrypt(key []byte, plaintext string) (string, error) {
	nonce := make([]byte, NonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return "", err
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	buf := gcm.Seal(nil, nonce, []byte(plaintext), nil)
	out := append(append([]byte{}, nonce...), buf...)
	return base64.StdEncoding.EncodeToString(out), nil
}

func Decrypt(key []byte, payload string) (string, error) {
	raw, err := base64.StdEncoding.DecodeString(payload)
	if err != nil || len(raw) < NonceLen+16 {
		return "", errors.New("invalid ciphertext")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	pt, err := gcm.Open(nil, raw[:NonceLen], raw[NonceLen:], nil)
	if err != nil {
		return "", err
	}
	return string(pt), nil
}

// ---------- 组织 DEK 惰性缓存 ----------

type DEKCache struct {
	mu   sync.Mutex
	deks map[string][]byte // orgID -> DEK(32B)
}

func NewDEKCache() *DEKCache { return &DEKCache{deks: map[string][]byte{}} }

func (c *DEKCache) Get(masterKey []byte, orgID, dekEncrypted string) ([]byte, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if dek, ok := c.deks[orgID]; ok {
		return dek, nil
	}
	hexDEK, err := Decrypt(masterKey, dekEncrypted)
	if err != nil {
		return nil, fmt.Errorf("unwrap org DEK: %w", err)
	}
	dek, err := hex.DecodeString(hexDEK)
	if err != nil || len(dek) != KeyLen {
		return nil, errors.New("invalid org DEK")
	}
	c.deks[orgID] = dek
	return dek, nil
}

// ---------- 密码哈希：Argon2id（新）+ scrypt（旧校验/迁移） ----------

type Params struct {
	Memory  uint32
	Time    uint32
	Threads uint8
	SaltLen int
	KeyLen  uint32
}

var Argon2Params = Params{Memory: 64 * 1024, Time: 3, Threads: 4, SaltLen: 16, KeyLen: 32}

func HashPassword(password string) (string, error) {
	salt := make([]byte, Argon2Params.SaltLen)
	if _, err := rand.Read(salt); err != nil {
		return "", err
	}
	hash := argon2.IDKey([]byte(password), salt,
		Argon2Params.Time, Argon2Params.Memory, Argon2Params.Threads, Argon2Params.KeyLen)
	return fmt.Sprintf("argon2id$v=19$m=%d,t=%d,p=%d$%s$%s",
		Argon2Params.Memory, Argon2Params.Time, Argon2Params.Threads,
		hex.EncodeToString(salt), hex.EncodeToString(hash)), nil
}

// CheckPassword 返回密码是否正确、以及存储哈希是否为旧 scrypt（需要重哈希）
func CheckPassword(password, stored string) (ok bool, needsRehash bool, err error) {
	parts := strings.Split(stored, "$")
	if len(parts) < 2 {
		return false, false, errors.New("malformed password hash")
	}
	switch parts[0] {
	case "argon2id":
		// argon2id$v=19$m=65536,t=3,p=4$salt$hash
		if len(parts) != 6 {
			return false, false, errors.New("malformed argon2id hash")
		}
		var m, t uint32
		var p uint8
		if _, err := fmt.Sscanf(parts[3], "m=%d,t=%d,p=%d", &m, &t, &p); err != nil {
			return false, false, err
		}
		salt, err := hex.DecodeString(parts[4])
		if err != nil {
			return false, false, err
		}
		want, err := hex.DecodeString(parts[5])
		if err != nil {
			return false, false, err
		}
		hash := argon2.IDKey([]byte(password), salt, t, m, p, uint32(len(want)))
		return subtle.ConstantTimeCompare(hash, want) == 1, false, nil
	case "scrypt":
		// scrypt$N$r$p$salt$hash（Node MVP 遗留格式）
		if len(parts) != 6 {
			return false, false, errors.New("malformed scrypt hash")
		}
		var N int
		var r, p int
		if _, err := fmt.Sscanf(parts[1], "%d", &N); err != nil {
			return false, false, err
		}
		if _, err := fmt.Sscanf(parts[2], "%d", &r); err != nil {
			return false, false, err
		}
		if _, err := fmt.Sscanf(parts[3], "%d", &p); err != nil {
			return false, false, err
		}
		salt, err := hex.DecodeString(parts[4])
		if err != nil {
			return false, false, err
		}
		want, err := hex.DecodeString(parts[5])
		if err != nil {
			return false, false, err
		}
		hash, err := scrypt.Key([]byte(password), salt, N, r, p, len(want))
		if err != nil {
			return false, false, err
		}
		return subtle.ConstantTimeCompare(hash, want) == 1, true, nil
	default:
		return false, false, errors.New("unknown hash algorithm")
	}
}

// ---------- JWT (HS256) ----------

func b64u(b []byte) string { return base64.RawURLEncoding.EncodeToString(b) }

func SignJWT(subject, email, secret string, ttl time.Duration) (string, error) {
	header := b64u([]byte(`{"alg":"HS256","typ":"JWT"}`))
	now := time.Now()
	body := b64u([]byte(fmt.Sprintf(`{"sub":%q,"email":%q,"iat":%d,"exp":%d}`,
		subject, email, now.Unix(), now.Add(ttl).Unix())))
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(header + "." + body))
	return header + "." + body + "." + b64u(mac.Sum(nil)), nil
}

func VerifyJWT(token, secret string) (sub string, err error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return "", errors.New("malformed token")
	}
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(parts[0] + "." + parts[1]))
	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil || subtle.ConstantTimeCompare(mac.Sum(nil), sig) != 1 {
		return "", errors.New("bad signature")
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return "", err
	}
	var claims struct {
		Sub string `json:"sub"`
		Exp int64  `json:"exp"`
	}
	if err := json.Unmarshal(payload, &claims); err != nil {
		return "", err
	}
	if claims.Exp < time.Now().Unix() {
		return "", errors.New("token expired")
	}
	return claims.Sub, nil
}

func SHA256(s string) string {
	sum := sha256.Sum256([]byte(s))
	return hex.EncodeToString(sum[:])
}

func NewID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
